pkg/ipam/allocator/multipool/allocator.go

```diff
@@ -104,3 +105,38 @@ func (a *Allocator) Init(ctx context.Context) (err error) {
 func (a *Allocator) Start(ctx context.Context, getterUpdater ipam.CiliumNodeGetterUpdater) (allocator.NodeEventHandler, error) {
 	return NewNodeHandler(a.poolAlloc, getterUpdater), nil
 }
+
+func (a *Allocator) UpsertPool(ctx context.Context, pool *cilium_v2alpha1.CiliumIPPool) error {
+	log.WithFields(logrus.Fields{
+		"pool-name":      pool.Name,
+		"ipv4-cidrs":     pool.Spec.IPv4.CIDRs,
+		"ipv4-mask-size": pool.Spec.IPv4.MaskSize,
+		"ipv6-cidrs":     pool.Spec.IPv6.CIDRs,
+		"ipv6-mask-size": pool.Spec.IPv6.MaskSize,
+	}).Debug("upserting pool")
+
+	ipv4CIDRs := make([]string, len(pool.Spec.IPv4.CIDRs))
+	for i, cidr := range pool.Spec.IPv4.CIDRs {
+		ipv4CIDRs[i] = string(cidr)
+	}
+	ipv6CIDRs := make([]string, len(pool.Spec.IPv6.CIDRs))
+	for i, cidr := range pool.Spec.IPv6.CIDRs {
+		ipv6CIDRs[i] = string(cidr)
+	}
+
+	return a.poolAlloc.UpsertPool(
```

Comment:
> Not sure how to best address that. Maybe we need to introduce a mutex to `PoolAllocator` itself?

Yes, looks like the cleanest solution currently is to introduce a mutex to `PoolAllocator`. For now I've added them to `PoolAllocator`'s exported API at the top level, i.e. it protects all of its fields (`pools`, `nodes`, `ready`). I'd assume this is fine for now since these operations aren't expected to be called frequently. If needed, we could refine the scope of the mutex as a follow-up.